File: ruby/classifier/perceptron.rb
require_relative '../classifier.rb'

module MachineLearning
  class Perceptron < Classifier
    def initialize(rho = 0.5, weight = nil)
      @rho = rho
      @weight = weight
    end

    def learn(train_data, train_class)
      bias_train_data = add_bias(train_data)
      if @weight
        weight = @weight
      else
        weight = initialize_weight(bias_train_data[0].size, train_class.uniq.size)
      end

      classify_correctly_all = false
      until classify_correctly_all
        classify_correctly_all = true
        bias_train_data.each_with_index do |data, i|
          discriminant_results = weight.map do |w|
            discriminant_function(data, w)
          end

          classification_class = discriminant_results.index(discriminant_results.max)
          actuals_class = train_class.uniq.sort.index(train_class[i])
          unless classification_class == actuals_class
            classify_correctly_all = false
            weight = update_weight(weight, data, @rho, classification_class, actuals_class)
          end
        end
      end
      @weight = weight
    end

    def clasify(test_data)
      bias_test_data = add_bias(test_data)
      bias_test_data.each_with_index do |data, i|
        puts "data: #{data.to_s}"
        print "discriminant_function_results: "
        r = @weight.map do |w|
          discriminant_function(data, w)
        end.to_s
        puts r
        puts
      end
    end

    private
    def add_bias(data)
      data.map { |d| [1.0] + d }
    end

    def initialize_weight(dimension, class_size)
      Array.new(class_size).map do |elem|
        Array.new(dimension).map { rand }
      end
    end

    def discriminant_function(data, weight)
      data.zip(weight).reduce(0) { |sum, arr| sum + arr[0] * arr[1] }
    end

    def update_weight(weight, data, rho, classification_class, actuals_class)
      weight[actuals_class] = weight[actuals_class].zip(data.map { |i| rho * i }).map { |i| i[0] + i[1] }
      weight[classification_class] = weight[classification_class].zip(data.map { |i| rho * i }).map { |i| i[0] - i[1] }
      weight
    end
  end
end

File: scala/src/main/scala/machinelearning/classifier/Perceptron.scala
package jp.mydns._2gmon.machinelearning.classifier {
  class Perceptron(val rho : Double = 0.5, val weight : List[List[Double]] = Nil) extends Classifier {
    def discriminantFunction(data : List[Double], weight : List[Double]) : Double = {
      data.zip(weight).map(a => a._1 * a._2).fold(0.0)((a, b) => a + b)
    }

    def learn(trainData : List[List[Double]], trainClass : List[Int]) : Perceptron = {
      val biasTrainData = trainData.map(1.0 :: _)
      def initializeWeight(classNum : Int, dimension: Int) : List[List[Double]] = {
        (1 to classNum).map(_ => (1 to dimension).map(_ => scala.util.Random.nextDouble).toList).toList
      }

      def perceptronStep(data : List[Double], weightList : List[List[Double]]) : List[List[Double]] = {
        def updateWeight(classificationClass : Int, actualsClass : Int) : List[List[Double]] = {
          weightList.map(weight =>
              if (weight == weightList(classificationClass))
                weight.zip(data.map(_ * rho)).map(a => a._1 - a._2)
              else if (weight == weightList(actualsClass))
                weight.zip(data.map(_ * rho)).map(a => a._1 + a._2)
              else
                weight
              )
        }

        val discriminantResult = weightList.map(weight => discriminantFunction(data, weight));
        val classificationClass = discriminantResult.indexOf(discriminantResult.max);
        val actualsClass = trainClass.distinct.sorted.indexOf(trainClass(biasTrainData.indexOf(data)));

        if (classificationClass == actualsClass)
          weightList
        else
          updateWeight(classificationClass, actualsClass)
      }

      def processAllData(trainData : List[List[Double]], weightList : List[List[Double]]) : List[List[Double]] = {
        if (trainData == List())
          weightList
        else
          processAllData(trainData.tail, perceptronStep(trainData.head, weightList))
      }

      def convergenceTest(weight : List[List[Double]], updatedWeight : List[List[Double]]) : List[List[Double]] = {
        if (weight == updatedWeight)
          updatedWeight
        else
          convergenceTest(updatedWeight, processAllData(biasTrainData, updatedWeight))
      }

      val initialWeight =
        if (weight == Nil)
          initializeWeight(trainClass.distinct.length, biasTrainData.head.length)
        else
          weight
      new Perceptron(rho, convergenceTest(List(List()), initialWeight))
    }

    def classify(trainData : List[List[Double]]) : List[Int] = {
      val biasTrainData = trainData.map(1.0 :: _)
      (for (data <- biasTrainData) yield {
        val discriminantResult = weight.map(w => discriminantFunction(data, w))
        discriminantResult.indexOf(discriminantResult.max)
      }).toList
    }
  }
}
